website/versioned_docs/version-v1.x/circuits.md

```diff
@@ -7,45 +7,407 @@ sidebar_position: 7
 
-Each circuit is parameterised and it is important to set the right parameters
-to your use case. For example, if you want to support up to 3125 messages, the message tree depth parameter should be set to `5` (as $5^5 = 3125$).
+Each circuit is parameterised and it is important to set the right parameters that matches your use case. For example, if you want to support up to 3125 messages, the message tree depth parameter should be set to `5` (as $5^5 = 3125$).
 
-Next, navigate to the `cli/` directory and edit `zkeys.config.yml`.
+## Background
 
-This config file defines the parameters required for MACI's circuits.
+### zk-SNARKs
 
-### Message processing
+MACI uses zk-SNARKs to prove that the coordinator has correctly processed a batch of messages and that the votes have been tallied correctly. zk-SNARKs are a type of zero-knowledge proof which allows a prover to prove that they know a secret without revealing the secret itself. In MACI, the prover is the coordinator. Users, can at any point after the proof generation, verify that the process was done correctly.
+
+### Circom
+
+MACI's circuits are written using [Circom](https://docs.circom.io/), a domain-specific language (DSL) used to write zk-SNARK circuits. Circom syntax resembles JavaScript, and it currently is one of the most popular DSL in use by zk developers. Please refer to their documentation to learn more about the language.
+
+### Proving system
+
+MACI uses [GROTH16](https://eprint.iacr.org/2016/260.pdf) as its proving system. GROTH16 is a zk-SNARK proving system that allows for the generation of proofs that are small and fast to verify.
+
+## How are the circuits used
+
+The circuits are used by the coordinator (the prover) to prove that they have correctly processed a batch of messages and tallied the votes correctly. This happens after a Poll has completed, and the coordinator has merged the state and message trees. The coordinator then generates a proof for each batch of messages, and submits them to the contract. The contract then verifies the proofs and updates the commitments on chain.
+
+> Please note that in order to function, the circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.
+
+## How do the Circuits fit in a voting round
+
+![flow](/img/circuits/processingAfterPollEnds.svg)
+
+## How the circuits work
+
+### Message processing (processMessages)
+
+This circuit allows the coordinator to prove that they have correctly processed each message in reverse order, in a consecutive batch of 5 ^ msgBatchDepth messages to the respective state leaf within the state tree. Coordinators would use this circuit to prove correct execution at the end of each Poll. The pre-requisites for this are:
```

Comment:
I think worth explaining why this happens in batches (because of gas limit, correct?)

Reply:
Well AFAIU we do it in batches because of circuits' constraints. If we were to process all messages in one go, the constraints would be so many that it would take forever and the circuit's artifacts (zkeys, r1cs, etc) would be MASSIVE